energy-integration/mex-governance/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+        }
+
+        if remaining_payment.amount > BigUint::zero() {
+            let caller = self.blockchain().get_caller();
+            self.send().direct_esdt(
+                &caller,
+                &remaining_payment.token_identifier,
+                remaining_payment.token_nonce,
+                &remaining_payment.amount,
+            );
+        }
+    }
+
+    #[endpoint(claimIncentive)]
+    fn claim_incentive(&self, week: Week) {
+        let current_week = self.get_current_week();
+        require!(week <= current_week, INVALID_INCENTIVE_WEEK);
+
+        let caller = self.blockchain().get_caller();
+        let user_id = self.user_ids().get_id_non_zero(&caller);
+        let incentive_token = self.incentive_token().get();
+
+        let user_votes = self.user_votes_in_week(user_id, week).get();
+        let mut user_payments = ManagedVec::new();
+        for user_vote in user_votes.iter() {
+            let farm_id = user_vote.farm_id;
+
+            let total_farm_incentive = self.farm_incentive_for_week(farm_id, week).get();
+            if total_farm_incentive > BigUint::zero() {
+                let total_farm_vote = self.farm_votes_for_week(farm_id, week).get();
+                let user_incentive =
+                    &total_farm_incentive * &user_vote.vote_amount / &total_farm_vote;
+
+                user_payments.push(EsdtTokenPayment::new(
+                    incentive_token.clone(),
+                    0,
+                    user_incentive,
+                ));
+            }
+        }
+
+        if user_payments.len() > 0 {
+            self.send().direct_multi(&caller, &user_payments);
+        }
+    }
+
+    fn advance_week_if_needed(&self, voting_week: usize) {
+        let saved_week = self.voting_week().get();
+        if saved_week < voting_week {
+            self.voting_week().set(voting_week);
+
+            let emission_rate = self.reference_emission_rate().get();
+            self.emission_rate_for_week(voting_week).set(emission_rate);
+        }
+    }
+
+    #[endpoint(setFarmEmissions)]
+    fn set_farm_emissions(&self) {
+        let current_week = self.get_current_week();
+        let emission_rate = self.emission_rate_for_week(current_week).get();
+        let total_votes = self.total_energy_voted(current_week).get();
+
+        for farm_id in self.voted_farms_for_week(current_week).iter() {
+            let farm_address_opt = self.farm_ids().get_address(farm_id);
+            require!(farm_address_opt.is_some(), FARM_NOT_FOUND);
+
+            let farm_address = unsafe { farm_address_opt.unwrap_unchecked() };
+
+            let farm_votes = self.farm_votes_for_week(farm_id, current_week).get();
+
+            let farm_emission = &emission_rate * &farm_votes / &total_votes;
+            self.farm_proxy(farm_address)
+                .set_per_block_rewards_endpoint(farm_emission)
+                .execute_on_dest_context::<()>();
+        }
+    }
```

Comment:
Own file. Also, check `total_votes > 0`. Sure, it's a very unlikely case, but what if we pause this contract for a week or something? Or if nobody votes for next week, but votes for week 2 from now?

Reply:
Moved to a new file. As for the 0 check, it is normal so that if no one voted for a farm in a certain week, to set it's emission rate to 0. As all farms generate rewards based on a single total amount.